File: spring_java/src/test/java/dev/pollito/spring_java/sakila/auth/adapter/in/rest/AuthRestControllerMockMvcTest.java
package dev.pollito.spring_java.sakila.auth.adapter.in.rest;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.when;
import static org.springframework.http.MediaType.APPLICATION_JSON;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

import dev.pollito.spring_java.config.security.userdetails.SakilaUserDetails;
import dev.pollito.spring_java.sakila.auth.domain.port.in.AuthUseCases;
import dev.pollito.spring_java.sakila.staff.domain.model.Staff;
import java.util.stream.Stream;
import org.jspecify.annotations.NonNull;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.MethodSource;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.webmvc.test.autoconfigure.AutoConfigureMockMvc;
import org.springframework.boot.webmvc.test.autoconfigure.WebMvcTest;
import org.springframework.context.annotation.Import;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.test.context.bean.override.mockito.MockitoBean;
import org.springframework.test.context.bean.override.mockito.MockitoSpyBean;
import org.springframework.test.web.servlet.MockMvc;

@WebMvcTest(AuthRestController.class)
@AutoConfigureMockMvc(addFilters = false)
@Import({AuthRestMapperImpl.class})
class AuthRestControllerMockMvcTest {

  @Autowired private MockMvc mockMvc;

  @MockitoBean private AuthUseCases authUseCases;

  @MockitoSpyBean private AuthRestMapper authRestMapper;

  @MockitoBean private UserDetailsService userDetailsService;

  @Test
  void loginReturnsToken() throws Exception {
    when(authUseCases.authenticate(any(), any())).thenReturn("jwt-token-123");

    mockMvc
        .perform(
            post("/api/auth/login")
                .contentType(APPLICATION_JSON)
                .content("{\"username\":\"Mike\",\"password\":\"password\"}")
                .accept(APPLICATION_JSON))
        .andExpect(status().isOk())
        .andExpect(jsonPath("$.data.token").value("jwt-token-123"));
  }

  static @NonNull Stream<SakilaUserDetails> getCurrentUserDetailsReturnsUserDetailsArgs() {
    return Stream.of(
        new SakilaUserDetails(
            Staff.builder().id(1).firstName("Mike").active(true).username("Mike").build()),
        null);
  }

  @ParameterizedTest
  @MethodSource("getCurrentUserDetailsReturnsUserDetailsArgs")
  void getCurrentUserDetailsReturnsUserDetails(SakilaUserDetails sakilaUserDetails)
      throws Exception {
    when(authUseCases.getCurrentUser()).thenReturn(sakilaUserDetails);

    var resultActions =
        mockMvc.perform(get("/api/auth/me").accept(APPLICATION_JSON)).andExpect(status().isOk());

    if (sakilaUserDetails != null) {
      resultActions
          .andExpect(jsonPath("$.data.username").value("Mike"))
          .andExpect(jsonPath("$.data.accountNonLocked").value(true))
          .andExpect(jsonPath("$.data.authorities[0]").value("ROLE_STAFF"))
          .andExpect(jsonPath("$.data.staff.username").value("Mike"));
    }
  }
}

File: spring_kotlin/src/test/kotlin/dev/pollito/spring_kotlin/sakila/auth/adapter/in/rest/AuthRestControllerMockMvcTest.kt
package dev.pollito.spring_kotlin.sakila.auth.adapter.`in`.rest

import com.ninjasquad.springmockk.MockkBean
import dev.pollito.spring_kotlin.config.security.userdetails.SakilaUserDetails
import dev.pollito.spring_kotlin.config.web.ControllerAdvice
import dev.pollito.spring_kotlin.sakila.auth.domain.port.`in`.AuthUseCases
import dev.pollito.spring_kotlin.sakila.staff.domain.model.Staff
import io.mockk.every
import kotlin.test.Test
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.CsvSource
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.webmvc.test.autoconfigure.WebMvcTest
import org.springframework.context.annotation.Import
import org.springframework.http.MediaType
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.get
import org.springframework.test.web.servlet.post

@WebMvcTest(AuthRestController::class)
@Import(ControllerAdvice::class, AuthRestMapperImpl::class)
class AuthRestControllerMockMvcTest {

  @MockkBean private lateinit var authUseCases: AuthUseCases
  @MockkBean private lateinit var userDetailsService: UserDetailsService
  @Autowired private lateinit var mockMvc: MockMvc

  @Test
  fun `login returns token`() {
    every { authUseCases.authenticate("Mike", "1234") } returns "jwt-token"

    mockMvc
        .post("/api/auth/login") {
          contentType = MediaType.APPLICATION_JSON
          content = "{\"username\":\"Mike\",\"password\":\"1234\"}"
        }
        .andExpect {
          status { isOk() }
          jsonPath("$.data.token") { value("jwt-token") }
        }
  }

  @ParameterizedTest
  @CsvSource("Mike,true", "null,false")
  fun `getCurrentUserDetails returns user details`(username: String, active: Boolean) {
    val staff =
        if (username == "null") null
        else
            Staff(
                id = 1,
                firstName = "Mike",
                lastName = "Hillyer",
                username = username,
                password = "password",
                email = "Mike.Hillyer@sakilastaff.com",
                active = active,
            )
    val userDetails = staff?.let { SakilaUserDetails(it) }
    if (userDetails != null) {
      every { authUseCases.getCurrentUser() } returns userDetails

      mockMvc
          .get("/api/auth/me") { accept = MediaType.APPLICATION_JSON }
          .andExpect {
            status { isOk() }
            jsonPath("$.data.username") { value(username) }
            jsonPath("$.data.accountNonLocked") { value(active) }
            jsonPath("$.data.authorities[0]") { value("ROLE_STAFF") }
            jsonPath("$.data.staff.username") { value(username) }
          }
    }
  }
}
